### PR TITLE
Remove binding_times_and_modes map from Aliases

### PR DESCRIPTION
**Pull request by mshinwell**
For compilation of `typecore.ml` at `-O3` (using the stage2 compiler) this reduces the number of instructions executed during compilation by about 1.5%.  CPU cycles went down by 0.7%.

### CODE REVIEW

File: middle_end/flambda2/types/env/aliases.ml [hunk 1]
```diff
@@ -427,42 +431,61 @@ let name_defined_earlier ~binding_time_resolver t alias ~than =
     else
       let alias_comp_unit = Name.compilation_unit alias in
       let than_comp_unit = Name.compilation_unit than in
+      let _current_comp_unit = Compilation_unit.get_current_exn () in
```
Comment (lthls) [inline middle_end/flambda2/types/env/aliases.ml:434]: Unused binding

**Review by lthls — APPROVED** [review]
The code looks correct.
One possible improvement (in terms of readability) would be to add `Binding_time.With_name_mode.symbols` (and `imported_variables`, and `consts_and_discriminants`) as the name mode is always the same for these.

**Comment by mshinwell** [comment]
> The code looks correct. One possible improvement (in terms of readability) would be to add `Binding_time.With_name_mode.symbols` (and `imported_variables`, and `consts_and_discriminants`) as the name mode is always the same for these.

I've made this improvement, which looks good.  I've also renamed `consts_and_discriminants` to just `consts` ("discriminants" was an old Flambda 2 concept for switch scrutinees).  I also removed the `lazy` from `initial_symbol_type` in `Typing_env` as it is no longer necessary.

**Comment by mshinwell** [comment]
@chambart Please catch up on this when you can.

**Pull request merged**